snakemake_executor_plugin_slurm_jobstep/__init__.py

```diff
@@ -53,47 +54,54 @@ def run_job(self, job: JobExecutorInterface):
         # snakemake_interface_executor_plugins.executors.base.SubmittedJobInfo.
 
         jobsteps = dict()
-        # TODO revisit special handling for group job levels via srun at a later stage
-        # if job.is_group():
-
-        #     def get_call(level_job, aux=""):
-        #         # we need this calculation, because of srun's greediness and
-        #         # SLURM's limits: it is not able to limit the memory if we divide the
-        #         # job per CPU by itself.
-
-        #         level_mem = level_job.resources.get("mem_mb")
-        #         if isinstance(level_mem, TBDString):
-        #             level_mem = 100
-
-        #         mem_per_cpu = max(level_mem // level_job.threads, 100)
-        #         exec_job = self.format_job_exec(level_job)
-
-        #         # Note: The '--exlusive' flag is a prevention for triggered job steps
-        #         #       within an allocation to oversubscribe within a given c-group.
-        #         #       As we are dealing only with smp software
-        #         #       the '--ntasks' is explicitly set to 1 by '-n1' per group job
-        #         #       (step).
-        #         return (
-        #             f"srun -J {job.groupid} --jobid {self.jobid}"
-        #             f" --mem-per-cpu {mem_per_cpu} -c {level_job.threads}"
-        #             f" --exclusive -n 1 {aux} {exec_job}"
-        #         )
-
-        #     for level in list(job.toposorted):
-        #         # we need to ensure order - any:
-        #         level_list = list(level)
-        #         for level_job in level_list[:-1]:
-        #             jobsteps[level_job] = subprocess.Popen(
-        #                 get_call(level_job), shell=True
-        #             )
-        #         # now: the last one
-        #         # this way, we ensure that level jobs depending on the current level
-        #         # get started
-        #         jobsteps[level_list[-1]] = subprocess.Popen(
-        #             get_call(level_list[-1], aux="--dependency=singleton"), shell=True
-        #         )
-
-        if "mpi" in job.resources.keys():
+
+        if job.is_group():
```

Comment:
This was deactivated by me because it neglected the fact that resources within a group will be mostly unknown beforehand, and further the topology does not really properly predict diverging runtimes of jobs, and it is therefore unnecessarily unflexible to group them by topology levels. However, I have not thought this through entirely and I might be wrong with my thoughts.

Reply:
Thing is: right now, group jobs do not work correctly. As far as I see it, there are two places which need editing:

- the submit executor; hence the CPU resources are `threads` (or `cpus_per_task`) times number of jobs, confined to a single node, unless we allow oversubscription
- the jobstep executor to handle one group job after the other (non-blocking, of course)

Reply:
ok, lets discuss this in a VC :-)